<a href="https://colab.research.google.com/github/zavaleta/UNT_2020/blob/master/Clase_unt_04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Curso de Data Science
## Universidad Nacional de Trujillo - Perú
### Enero 08 a 15 de 2020
#### Prof. Dr. Jorge Zavaleta
##### zavaleta.jorge@gmail.com

## Deep Learning (aprendizaje profundo)

El aprendizaje profundo (Deep Learning), tambien conocido como aprendizaje estruturado profundo es una rama del aprendizaje de máquina basado en un conjunto de algoritmos que intentan modelar abstracciones de alto nivel de datos usando un grafo profundo con varias camadas de procesamiento, compuesta de varias transformaciones lineares y no leneares. Una de las promesas del aprendizaje profundo es la subtitución de características hechas manualmente por algoritmos eficientes por un aprendizaje de caracteristicas supervisionadas o semisupervisionadas y extracción jerárquica de características.

![image.png](attachment:image.png)

El aprendizaje profundo es responsable por los ultimos avanzes en visión computacional, reconocimiento de voz, procesamiento de lenguaje natural (PLN) y reconocimiento de sonido. El aprendizaje profundo esta basado en el concepto de redes neurales artificiales, o sistemas computacionales que imitan la manera como el cerebro humano funciona.

# Funcionamiento del Deep Learning

Un modelo de aprendizaje profundo es projectado para analisar continuamente los datos con una estructura lógica semejante a la manera como un ser humano llegaria a conclusiones.

Para conseguir eso, los aplicativos de aprendizaje profundo usam una estructura en camadas de algoritmos llamada de red neural artificial.

El diseño de una RNA es inspirado en la red biologica del cerebro humano, llevando a un proceso de aprendizaje mucho mas capaz a los realizados por los modelos estandares del aprendizado de máquina.

## Ejemplos practicos

### Asistentes Virtuales

Los asistentes virtuales de proveedoresde servicios on-line usan el aperendizaje profundo para ajudar a entender su discurso y el lenguaje que los humanos usan para interajir entre ellos.
- Alexa - https://www.showmetech.com.br/como-moderniza-sua-casa-com-a-amazon-echo/
- Siri - https://www.apple.com/br/siri/
- Cortana do Windows

### Traducciones

Los algoritmos de aprendizaje profundo puede produzir automaticamente diversos tipos de idiomas. Herramienta poderosa para viajantes, personas de negocios y personas de gobierno que precisan de una traducción rápida y eficiente.

### Visión para camiones de entrega sin motorista, drones y carros autonomos

La manera como un carro autonomo comprende las realidades das carreteras y como responder a ellas, sea una señal para parar, una bola en la calle, u otro carro, es atravez de algoritmos de aprendizaje profundo. Cuantos mas datos estos algoritmos reciben, mejor ellos son capazes de actuar como seres humanos en su procesamiento de informaciones. 

### Servicio de Chatbots y bots

Los servicios de **Chatbots** y **robots** proporcionan atendimiento a los clientes para muchas empresas son capazes de responder de una manera inteligente y util a una cantidad creciente de preguntas auditivas y de texto, gracias al aprendizaje profundo.

### Colorización de imagenes

Transformar imagenes de negro y blanco en colores que es una tarea compleja hecho a mano. Actualmente, los algoritmos de aprendizaje profunda son capazes de usar el contexto y los objetos en las imagenes para colorirlos, basicamente, para recrear la imagen en negro y blanco en colores. Los resultados son impresionantes y precisos.

### Reconocimiento facial

El aprendizaje profundo está siendo utilizando para el reconocimiento facial, por razones de seguridad, para etiquetar
a las personas en las publicaciones de Facebook y es posible que podamos pagar los artículos en una tienda simplemente usando nuestras caras en un futuro cercano. Los desafíos para los algoritmos de aprendizaje profundo para el reconocimiento facial es saber que son la misma persona incluso cuando cambian su peinado, crecen o se afeitan el cabello, o si la imagen capturada es mala debido a una iluminación deficiente o una obstrucción.

### Medicina y Farmacéutica

Desde diagnósticos de enfermedades y tumores hasta medicamentos personalizados diseñados específicamente para el genoma de un individuo, el aprendizaje médico en profundidad tiene la atención de muchas de las compañías farmacéuticas y médicas más grandes.

### Compras y entretenimientos personalizados

Você já se perguntou como a Netflix apresenta sugestões para o que você deve assistir a seguir? Ou onde a Amazon surge com idéias para o que você deve comprar em seguida e essas sugestões são exatamente o que você precisa, mas nunca soube disso antes? Sim, é algoritmo de aprendizagem profunda no trabalho.

¿Alguna vez se preguntó cómo Netflix ofrece sugerencias sobre qué ver a continuación? ¿O dónde Amazon presenta ideas sobre qué comprar a continuación y estas sugerencias son exactamente lo que necesita, pero nunca lo supo antes? Sí, es un algoritmo de aprendizaje profundo en el trabajo.

# Herramientas para Aprendizado Profundo

![image.png](attachment:image.png)

## Keras
Keras es una API de Redes Neurales de alto nivel escrito en Python y puede executarse sobre TensorFlow, Theano y CNTK.

Se deve usar el Keras cuando sea necesario una biblioteca de aprendizaje profundo que:

- Permite la creación de prototipos fácil y rápida (a través de la facilidad de uso, la modularidad y la extensibilidad).
- Admite redes convolucionales y redes recurrentes, así como combinaciones de las dos.
- Se ejecuta sin problemas en CPU y GPU.

## TensorFlow

TensorFlow es una biblioteca de aprendizaje automático de código abierto aplicable a una amplia variedad de tareas. Es un sistema para crear y entrenar redes neuronales para detectar y descifrar patrones y correlaciones, de forma análoga a la forma en que los humanos aprenden y razonan.

### Theano

Theano es una biblioteca de Python y un compilador de optimización para manipular y evaluar expresiones matemáticas, especialmente aquellas con valor de matriz. 

En Theano, los cálculos se expresan utilizando la sintaxis NumPy y se compilan para funcionar de manera eficiente en arquitecturas de CPU o GPU.

### CNTK

Microsoft Cognitive Toolkit, anteriormente conocido como **CNTK** y a veces llamado The Microsoft Cognitive Toolkit, es un marco de aprendizaje profundo desarrollado por Microsoft Research.

## Redes convulucionales - CNN

![image.png](attachment:image.png)

La CNN es un tipo de Red Neuronal Artificial con aprendizaje supervisado que procesa sus capas imitando al cortex visual del ojo humano para identificar distintas características en las entradas que en definitiva hacen que pueda identificar objetos y «ver». Para ello, la CNN contiene varias capas ocultas especializadas y con una jerarquía: esto quiere decir que las primeras capas pueden detectar lineas, curvas y se van especializando hasta llegar a capas más profundas que reconocen formas complejas como un rostro o la silueta de un animal.

![image.png](attachment:image.png)

![image.png](attachment:image.png)

## Ejemplo 1 - Keras

In [0]:
## conda install -c conda-forge keras

In [0]:
# from keras.layers import Activation, Dense
# from keras.models import Sequential
# model = Sequential()
# model.add(Dense(12, input_dim=8, kernel_initializer='random_uniform'))

In [0]:
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility

Using TensorFlow backend.


In [0]:
# network and training
NB_EPOCH = 200
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # SGD optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION

In [0]:
# data: shuffled and split between train and test sets
#
(X_train, y_train), (X_test, y_test) = mnist.load_data()
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784
#X
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
#X
X_train /= 255
X_test /= 255
print(X_train.shape[0], 'ejemplo de entrenamiento')
print(X_test.shape[0], 'ejemplo de teste')
# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)

60000 ejemplo de entrenamiento
10000 ejemplo de teste


In [0]:
# 10 outputs
# final stage is softmax
model = Sequential()
model.add(Dense(NB_CLASSES, input_shape=(RESHAPED,)))
model.add(Activation('softmax'))
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 10)                7850      
_________________________________________________________________
activation_1 (Activation)    (None, 10)                0         
Total params: 7,850
Trainable params: 7,850
Non-trainable params: 0
_________________________________________________________________


In [0]:
model.compile(loss='categorical_crossentropy', optimizer=OPTIMIZER, metrics=['accuracy'])

In [0]:
history = model.fit(X_train, Y_train,
batch_size=BATCH_SIZE, epochs=NB_EPOCH,
verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

Train on 48000 samples, validate on 12000 samples
Epoch 1/200
48000/48000 [==============================] - 6s 124us/step - loss: 1.3633 - accuracy: 0.6796 - val_loss: 0.8904 - val_accuracy: 0.8246
Epoch 2/200
48000/48000 [==============================] - 2s 46us/step - loss: 0.7913 - accuracy: 0.8272 - val_loss: 0.6572 - val_accuracy: 0.8546
Epoch 3/200
48000/48000 [==============================] - 2s 47us/step - loss: 0.6436 - accuracy: 0.8497 - val_loss: 0.5625 - val_accuracy: 0.8681
Epoch 4/200
48000/48000 [==============================] - 2s 48us/step - loss: 0.5717 - accuracy: 0.8602 - val_loss: 0.5098 - val_accuracy: 0.8765
Epoch 5/200
48000/48000 [==============================] - 2s 47us/step - loss: 0.5276 - accuracy: 0.8677 - val_loss: 0.4758 - val_accuracy: 0.8826
Epoch 6/200
48000/48000 [==============================] - 2s 49us/step - loss: 0.4973 - accuracy: 0.8726 - val_loss: 0.4515 - val_accuracy: 0.8866
Epoch 7/200
48000/48000 [==============================] - 3s

Epoch 56/200
48000/48000 [==============================] - 2s 43us/step - loss: 0.3164 - accuracy: 0.9120 - val_loss: 0.3037 - val_accuracy: 0.9156
Epoch 57/200
48000/48000 [==============================] - 2s 42us/step - loss: 0.3157 - accuracy: 0.9128 - val_loss: 0.3034 - val_accuracy: 0.9152
Epoch 58/200
48000/48000 [==============================] - 2s 43us/step - loss: 0.3149 - accuracy: 0.9121 - val_loss: 0.3029 - val_accuracy: 0.9148
Epoch 59/200
48000/48000 [==============================] - 2s 43us/step - loss: 0.3143 - accuracy: 0.9128 - val_loss: 0.3022 - val_accuracy: 0.9151
Epoch 60/200
48000/48000 [==============================] - 2s 43us/step - loss: 0.3136 - accuracy: 0.9129 - val_loss: 0.3016 - val_accuracy: 0.9161
Epoch 61/200
48000/48000 [==============================] - 2s 42us/step - loss: 0.3130 - accuracy: 0.9132 - val_loss: 0.3011 - val_accuracy: 0.9158
Epoch 62/200
48000/48000 [==============================] - 2s 42us/step - loss: 0.3123 - accuracy: 0.9131

Epoch 111/200
48000/48000 [==============================] - 2s 47us/step - loss: 0.2922 - accuracy: 0.9183 - val_loss: 0.2858 - val_accuracy: 0.9209
Epoch 112/200
48000/48000 [==============================] - 2s 46us/step - loss: 0.2919 - accuracy: 0.9187 - val_loss: 0.2857 - val_accuracy: 0.9207
Epoch 113/200
48000/48000 [==============================] - 2s 45us/step - loss: 0.2916 - accuracy: 0.9189 - val_loss: 0.2855 - val_accuracy: 0.9210
Epoch 114/200
48000/48000 [==============================] - 2s 45us/step - loss: 0.2913 - accuracy: 0.9188 - val_loss: 0.2853 - val_accuracy: 0.9212
Epoch 115/200
48000/48000 [==============================] - 2s 45us/step - loss: 0.2911 - accuracy: 0.9189 - val_loss: 0.2853 - val_accuracy: 0.9206
Epoch 116/200
48000/48000 [==============================] - 2s 52us/step - loss: 0.2908 - accuracy: 0.9190 - val_loss: 0.2849 - val_accuracy: 0.9213
Epoch 117/200
48000/48000 [==============================] - 2s 50us/step - loss: 0.2906 - accuracy:

48000/48000 [==============================] - 3s 55us/step - loss: 0.2810 - accuracy: 0.9215 - val_loss: 0.2785 - val_accuracy: 0.9229
Epoch 166/200
48000/48000 [==============================] - 3s 60us/step - loss: 0.2808 - accuracy: 0.9215 - val_loss: 0.2785 - val_accuracy: 0.9225
Epoch 167/200
48000/48000 [==============================] - 2s 51us/step - loss: 0.2807 - accuracy: 0.9217 - val_loss: 0.2784 - val_accuracy: 0.9227
Epoch 168/200
48000/48000 [==============================] - 2s 46us/step - loss: 0.2805 - accuracy: 0.9219 - val_loss: 0.2783 - val_accuracy: 0.9227
Epoch 169/200
48000/48000 [==============================] - 3s 53us/step - loss: 0.2804 - accuracy: 0.9217 - val_loss: 0.2782 - val_accuracy: 0.9227
Epoch 170/200
48000/48000 [==============================] - 3s 61us/step - loss: 0.2802 - accuracy: 0.9217 - val_loss: 0.2781 - val_accuracy: 0.9226
Epoch 171/200
48000/48000 [==============================] - 3s 53us/step - loss: 0.2801 - accuracy: 0.9219 - val_

In [0]:
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Teste de escore:", score[0])
print('Teste precisión:', score[1])

10000/10000 [==============================] - 1s 64us/step
Teste de escore: 0.27738100505769253
Teste precisión: 0.9225000143051147


## Ejemplo 2 - Tensorflow

In [0]:
# Tensorflow
import numpy as np
import tensorflow as tf # tensorflow
from keras.layers import Activation, Dense
from keras.models import Sequential
#
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility

### Ativación

In [0]:
import tensorflow.compat.v1 as tfc
sess = tfc.InteractiveSession()
#tf.InteractiveSession()

### Define el tensorflow

In [0]:
a = tfc.zeros((2,2));
b = tfc.ones((2,2))
print('a =',a)
print('b =',b)
# matriz de 4 x 4
my_tensor = tfc.random_uniform((4, 4), 0, 1)
print('MTF = ',my_tensor)
mtf = tfc.random_uniform((8, 4), 0, 1)
print('MTF = ',mtf)

a = tf.Tensor(
[[0. 0.]
 [0. 0.]], shape=(2, 2), dtype=float32)
b = tf.Tensor(
[[1. 1.]
 [1. 1.]], shape=(2, 2), dtype=float32)
MTF =  tf.Tensor(
[[0.694355   0.6712259  0.05648172 0.155231  ]
 [0.51512754 0.05355239 0.949628   0.64937615]
 [0.29666936 0.18530798 0.1556176  0.2856332 ]
 [0.72291243 0.30500567 0.63851273 0.7518152 ]], shape=(4, 4), dtype=float32)
MTF =  tf.Tensor(
[[0.76596344 0.62523234 0.1597569  0.13259685]
 [0.36473525 0.02367711 0.781121   0.88834417]
 [0.51170456 0.25561237 0.3578694  0.16334832]
 [0.82001495 0.23723257 0.73076284 0.25448155]
 [0.78167367 0.84630203 0.91765857 0.93548715]
 [0.1883837  0.2622397  0.44603896 0.70673287]
 [0.3248142  0.62689984 0.44797873 0.6378888 ]
 [0.29057574 0.7102716  0.57152236 0.6982354 ]], shape=(8, 4), dtype=float32)


### Operaciones

In [0]:
# Sumade elementos de una matriz 2D (2D Tensor) 
r = tfc.reduce_sum(b,reduction_indices = 1)
print('R=',r)

R= tf.Tensor([2. 2.], shape=(2,), dtype=float32)


In [0]:
a.get_shape()

TensorShape([2, 2])

# Reconocimiento de Imagenes

## Objetivo

![image.png](attachment:image.png)

## Caracteristica?

![image.png](attachment:image.png)

## Clasificador

### clasificador lineal?
![image.png](attachment:image.png)

## Clasificador

![image.png](attachment:image.png)

## Conjunto de entrenamiento

![image.png](attachment:image.png)

## Reconocimiento de imagenes clásica

![image.png](attachment:image.png)

## Historia de los métodos de Visión Computacional

- Descriptores de colores, forma y textura (1970 - 2000)
- Escala invariante de caracteristicas (>1999)
- Histograma de gradientes (>2005)
- Bolsa de caracteristicas (>2004)
- Piramide Espacial de Coincidencias (>2006)

# Para verificar

![image.png](attachment:image.png)

imageNet Challenge: ~ 1.4 millones de imagenes, 1000 clases

## Para executar
https://www.tensorflow.org/tutorials/keras/classification